# Changelog

## 1.0.0 (2026-02-25)


### Features

* add constrain deeptools_coverage to only process bigwigs that match only sample name ([6d5edc2](https://github.com/aryazand/fastq-process-align/commit/6d5edc2e61651e4539c408480a82e87556373f8d))
* added bowtie2 mapper ([17fd091](https://github.com/aryazand/fastq-process-align/commit/17fd091a8eae6c54aa87d3f493976783fc59ba3e))
* added bwa-mem2 as mapper ([b2508b2](https://github.com/aryazand/fastq-process-align/commit/b2508b2cb566969321e1dd4632ddc365370eee62))
* added deeptools to produce bigwig cov ([04dbd94](https://github.com/aryazand/fastq-process-align/commit/04dbd94d48af4a1afe6ecab9c17169e18d78bbe2))
* added first draft of report for most important variants ([5cf5150](https://github.com/aryazand/fastq-process-align/commit/5cf515029d910949d6e7f91caa4477674a0d40bd))
* added freebayes as second variant caller ([d4d49b8](https://github.com/aryazand/fastq-process-align/commit/d4d49b8a60309e59aa60c49df51f377a7aa594ee))
* added genome plots to report ([c4acfb1](https://github.com/aryazand/fastq-process-align/commit/c4acfb164ca10cc6f50fd83559993bf2398f8b72))
* added license ([a4831f3](https://github.com/aryazand/fastq-process-align/commit/a4831f370001ac4708368954906497b00269f28c))
* added minimap2 as another mapper option ([99e95d1](https://github.com/aryazand/fastq-process-align/commit/99e95d1be49984cd061e894f996287d30cb2e3d5))
* added module to retrieve mapping stats, exposed params ([421001a](https://github.com/aryazand/fastq-process-align/commit/421001aca615a899c5ad653015571cda89ea77ae))
* added new modules for mapping ([82cacb3](https://github.com/aryazand/fastq-process-align/commit/82cacb36fc7bf67cbbff7a218d02dcc6fcbbc80c))
* added option to add overhang for circular chromosomes ([43e2099](https://github.com/aryazand/fastq-process-align/commit/43e2099443452207dc538b00e73b51d7cc10551f))
* added rule to collect consensus variants ([f883811](https://github.com/aryazand/fastq-process-align/commit/f883811997a54a3cfdcb42c2db98a71b6ca787b8))
* added schemas for sample sheet and config files ([87385b7](https://github.com/aryazand/fastq-process-align/commit/87385b71f9f8d645ce1b9080cfa533913aac1aef))
* added SnpEff part 1 ([52b4f1d](https://github.com/aryazand/fastq-process-align/commit/52b4f1dc7565547ecb50363a51883b62c5cb85cd))
* added SnpEff part 2 ([59c2e69](https://github.com/aryazand/fastq-process-align/commit/59c2e6973e5cf6795173e95c2191926b6ae8611b))
* added STAR as a mapper option ([9dd4fa8](https://github.com/aryazand/fastq-process-align/commit/9dd4fa8931be24a20e5147a557d73e558f55ef7a))
* added test data set ([2f917ca](https://github.com/aryazand/fastq-process-align/commit/2f917ca0681c9e14bac8629e4cc809ee586bc003))
* added variant effect prediction module ([9e9d940](https://github.com/aryazand/fastq-process-align/commit/9e9d940e2e6af376b58802ce8c47f7c53d0b7bdb))
* added VEP for variant prediction ([354548a](https://github.com/aryazand/fastq-process-align/commit/354548aab52363677e957001245d177a9f00cae9))
* adding mappers, various improvements ([66756d2](https://github.com/aryazand/fastq-process-align/commit/66756d252606005abbe6e312872a786addd51c9c))
* enable apptainer deployment ([f0198f6](https://github.com/aryazand/fastq-process-align/commit/f0198f624fee081398c60121b3e85e51ebe365a5))
* enable apptainer deployment ([138414b](https://github.com/aryazand/fastq-process-align/commit/138414b5063a91567a7de2dc0ad987da2cb08b64))
* fix: added option to choose varaint caller to config ([96f51d7](https://github.com/aryazand/fastq-process-align/commit/96f51d7f242a4221ad0d3805723b073b408c3271))
* initial commit ([822a07c](https://github.com/aryazand/fastq-process-align/commit/822a07c49727e72a91bf903b34fa4604f708695e))
* more detailed plots for report ([9cdc49a](https://github.com/aryazand/fastq-process-align/commit/9cdc49aba3422c258b5780cf5e4d73e8fae8994d))
* output metrics data from bowtie2 ([f70434b](https://github.com/aryazand/fastq-process-align/commit/f70434be20433360eec74f1342d3828d5db83919))
* perform overhang adjustment in one awk script ([073444c](https://github.com/aryazand/fastq-process-align/commit/073444cc2835a0ea1b6c846949920b7ab6e0ded4))
* transfer reads from artificial overhang to correct positions in bam files ([18dda98](https://github.com/aryazand/fastq-process-align/commit/18dda98e0c0fe8d30d9fb062957a0fdcd5e5193e))
* use cram instead of bam and use temp bam files ([003b71c](https://github.com/aryazand/fastq-process-align/commit/003b71c4b58e13961b9985007786447b34b37824))
* various improvements regarding input options, docs ([fc2fa4d](https://github.com/aryazand/fastq-process-align/commit/fc2fa4dc2608c840f012b3b8b0db276be450ba1b))
* various improvements to variant effect prediction + viz ([103ada1](https://github.com/aryazand/fastq-process-align/commit/103ada16e44fed984e793e460a7004b31f90b81a))


### Bug Fixes

* Add channel_priority to umi_tools.yml ([ca842c1](https://github.com/aryazand/fastq-process-align/commit/ca842c123ecedf07a986a0cdccb47dd4f429fbee))
* added dag as workflow overview ([24be9b4](https://github.com/aryazand/fastq-process-align/commit/24be9b4ef86425142f5464185a2eb6e4bb5fc80d))
* added missing argument input to get_genome ([532afee](https://github.com/aryazand/fastq-process-align/commit/532afeea3b91a37f55b0e0d09f86898881a21caf))
* added missing documentation ([b2156eb](https://github.com/aryazand/fastq-process-align/commit/b2156eb40057aa11e37a0aa8abc09a5f830fcbd2))
* added script to reformat GFF for VEP ([6d19814](https://github.com/aryazand/fastq-process-align/commit/6d19814a813ff0d5cc7f05f0de8b69d28bc2f6f9))
* adjusted resource allocation ([5e7d8a3](https://github.com/aryazand/fastq-process-align/commit/5e7d8a3ce1c31c3f2023beb8adb36dd0b6f1b14c))
* adjusted resource definitions ([8667d70](https://github.com/aryazand/fastq-process-align/commit/8667d70278cef36e6fc556a772e958fe85817f95))
* bug setting cores dynamically, use static number ([bbe6509](https://github.com/aryazand/fastq-process-align/commit/bbe6509361794ba3b4e0bff6e9e323297c647aaf))
* chunks with default messaging turned off ([0310f01](https://github.com/aryazand/fastq-process-align/commit/0310f0122b837e640738b7f64b3be43c0c2df48c))
* corrected dir for building Dockerfile ([d5e752a](https://github.com/aryazand/fastq-process-align/commit/d5e752aa1c7269dc70b62e1e641ae449ba6c15db))
* corrected dir for building Dockerfile ([79933c2](https://github.com/aryazand/fastq-process-align/commit/79933c2709ca0eb5fbe7e70a18ab42124bbc1a52))
* corrected urls to json schema ([d9456d9](https://github.com/aryazand/fastq-process-align/commit/d9456d994b11fc2ea08d3baa303167cb05ef6d8f))
* explicit link to fasta index ([5a7e895](https://github.com/aryazand/fastq-process-align/commit/5a7e89537b1cc4e6cd6e65ae1113f09871995ad6))
* expose extra args in config ([93d4500](https://github.com/aryazand/fastq-process-align/commit/93d4500a26594bd774b3a240f45e3eab64096eac))
* exposed parameters for some tools ([068d30d](https://github.com/aryazand/fastq-process-align/commit/068d30d74c7ff6eb62deb5fc169ba7bd49f7a3a8))
* exposed parameters for some tools ([55a0ac4](https://github.com/aryazand/fastq-process-align/commit/55a0ac4addcbffb75d260eb5fb1524d0db1c1ba3))
* fixed get_processed_fastq to result in the fastq file after adapter/quality trimming, not UMI extraction ([eebb9e7](https://github.com/aryazand/fastq-process-align/commit/eebb9e73cb797859d1b505ead0a4a0f72512ebc6))
* handling of input files in umi_tools_extract. Modified input_args lambda to accept output parameter and use full output paths directly in -S and --read2-out instead of relative filenames ([59a8f59](https://github.com/aryazand/fastq-process-align/commit/59a8f59b60221f5ebfcf0947376ee4c431ffb2f1))
* handling of input files in umi_tools_extract. Modified input_args lambda to accept output parameter and use full output paths directly in -S and --read2-out instead of relative filenames ([d092fbd](https://github.com/aryazand/fastq-process-align/commit/d092fbdf0db0f30f13a413ab966942430b94a664))
* how lambda input statement in bam_to_cram rule ([c4273ae](https://github.com/aryazand/fastq-process-align/commit/c4273aeb8987dcb3db54951e5a123be2092a5224))
* improve handling of circular chromosomes ([f47fcc1](https://github.com/aryazand/fastq-process-align/commit/f47fcc1f2a9b75fc3388b08d31047a9ed98b697c))
* index correct fasta ([4e1418b](https://github.com/aryazand/fastq-process-align/commit/4e1418b08c2d96a27c44ac27321f4bd45e4dda8c))
* linting ([2b8d7d0](https://github.com/aryazand/fastq-process-align/commit/2b8d7d06bb8b3874935fa902cc8f4ab19d96cee2))
* location of vep log file ([9d5d7a8](https://github.com/aryazand/fastq-process-align/commit/9d5d7a803ac15a97abf026cab8d1716362174f66))
* make multiqc recognize vep output ([a01343b](https://github.com/aryazand/fastq-process-align/commit/a01343b7dc40876012a9d4c45300e1077f1349e5))
* minor changes config settings ([aeeab0e](https://github.com/aryazand/fastq-process-align/commit/aeeab0e22f185b785684b61d1051d14e3276fc99))
* minor fixes for next release ([13a1e03](https://github.com/aryazand/fastq-process-align/commit/13a1e03b634ba6e1675135a9adf32bd9758fd184))
* minor formatting ([338d9ee](https://github.com/aryazand/fastq-process-align/commit/338d9eecfdd2be5da16015800d88550699718fe6))
* minor issues, formatting ([865e609](https://github.com/aryazand/fastq-process-align/commit/865e6094607048153f1902439744aa654711fcd1))
* more changes to convert NCBI to ensemble GFF format ([8a9a948](https://github.com/aryazand/fastq-process-align/commit/8a9a948004d30e4c7df10f7c654504c6d2c33c87))
* name output of rule add_overhang_for_circular_chromosomes ([ee7b4ca](https://github.com/aryazand/fastq-process-align/commit/ee7b4ca47dfa7b41304629aa66bbe7613879ac9e))
* paths to run wf in main dir ([c643bfd](https://github.com/aryazand/fastq-process-align/commit/c643bfd71e487b6802081363141f183dce75bb3f))
* paths to run wf in main dir ([7fecab2](https://github.com/aryazand/fastq-process-align/commit/7fecab2f516792f19db70f0c526005f81e04cea5))
* perform UMI extraction before quality and adapter trimming ([f2484fc](https://github.com/aryazand/fastq-process-align/commit/f2484fc3199ffe77c98436b83fcb7a32e9fbfff4))
* prevent report from failing when no variants were found + format ([538c772](https://github.com/aryazand/fastq-process-align/commit/538c7723f2a32cf25c7da1b41e607506df9f9813))
* prevent rule from crashing when no variants were found at all ([c7ec81c](https://github.com/aryazand/fastq-process-align/commit/c7ec81ce3b1fd8ee96f229c1528e9bedda8fcf65))
* properly reference genome to produce cram files ([d38fde2](https://github.com/aryazand/fastq-process-align/commit/d38fde23dea7a8d2b3ccd57455ce974d163f920c))
* remove default container from prior developer ([a846162](https://github.com/aryazand/fastq-process-align/commit/a84616203d4d7e94e03bfc29a6077f2022bd42d8))
* remove temp marker on alignment output ([648c8ba](https://github.com/aryazand/fastq-process-align/commit/648c8baaac6a571497b39aa2565c4f1b856351e1))
* removed extra reference to fastp log files ([8cbbd22](https://github.com/aryazand/fastq-process-align/commit/8cbbd22e891aae23fdba928d769f1c74028aff8c))
* removed phase attribute for exons and mRNAs ([25fabe2](https://github.com/aryazand/fastq-process-align/commit/25fabe20356f7b5d0edc70e779f274813c0c873c))
* removed unnecessary fastp_se rule ([232a784](https://github.com/aryazand/fastq-process-align/commit/232a784899a34e3c8127965c387ba9d8d29aa141))
* removed unnecessary rdata export ([aa6a5d5](https://github.com/aryazand/fastq-process-align/commit/aa6a5d5dc6535591b1fba01a0f6a83f82dcdadde))
* report chromosomes etc as tabs ([029c1df](https://github.com/aryazand/fastq-process-align/commit/029c1dfafe26354a1e3f773b5b3e29872920c99f))
* restrict number of variants for heatmaps ([f08d201](https://github.com/aryazand/fastq-process-align/commit/f08d2017dae6c1de1f11bb5f24728056c6b32023))
* tolerate empty genome input ([8f3d0cf](https://github.com/aryazand/fastq-process-align/commit/8f3d0cf9b1b920d870b26e7b60d33aee2fbb9f5a))
* update dag ([ce4c68c](https://github.com/aryazand/fastq-process-align/commit/ce4c68c9fc91d4e6d60ad0464c5de91773d57762))
* update gffread_gff to take in genome with overhangs if specified ([412fd4f](https://github.com/aryazand/fastq-process-align/commit/412fd4f32ee08d131f4374f732c39659ac036cc8))
* use empty array instead of empty string for no input ([927586d](https://github.com/aryazand/fastq-process-align/commit/927586d141dad666631dc38725b87394375f2e69))
* works now seamless with paired and single end input ([5844af8](https://github.com/aryazand/fastq-process-align/commit/5844af810006f0f5cc5bd847d2aaf8530c33a299))
